nandtool/nandtool.c

```diff
@@ -177,6 +177,91 @@ static int nandtool_check(void)
 }
 
 
+static int nandtool_dump(const char *outpath, unsigned int start, unsigned int nblocks, int oob)
+{
+	flashsrv_info_t *info;
+	int ret;
+	size_t blocksz, partsz, pagesz;
+	size_t addr, endaddr;
+	uint8_t *buf;
+	int outfd;
+
+	outfd = open(outpath, O_RDWR | O_CREAT);
+	if (outfd < 0) {
+		fprintf(stderr, "nandtool: Fail to open file: %s\n", outpath);
+		return -EINVAL;
+	}
+
+	info = flashmng_info(nandtool_common.oid);
+	if (info == NULL) {
+		close(outfd);
+		fprintf(stderr, "nandtool: Fail to retrieve partition info\n");
+		return -EFAULT;
+	}
+
+	if (!oob) {
+		blocksz = info->erasesz;
+		partsz = info->size;
+		pagesz = info->writesz;
+	}
+	else {
+		pagesz = info->metasz + info->writesz;
+		blocksz = (info->erasesz / info->writesz) * pagesz;
+		partsz = (info->size / info->erasesz) * blocksz;
+	}
+
+	addr = start * blocksz;
+
+	if (!oob) {
+		if (lseek(nandtool_common.fd, addr, SEEK_SET) < 0) {
+			fprintf(stderr, "nandtool: lseek failed\n");
+			close(outfd);
+			return -EINVAL;
+		}
+	}
+
+	buf = malloc(pagesz);
+	if (buf == NULL) {
+		close(outfd);
+		fprintf(stderr, "nandtool: Fail to allocate memory for buffer!\n");
+		return -ENOMEM;
+	}
+
+	if (nblocks == 0) {
+		/* Read everything from start to end of partition */
+		endaddr = partsz;
+	}
+	else {
+		endaddr = addr + nblocks * blocksz;
+	}
+
+	while (addr < endaddr) {
+		if (oob) {
+			ret = flashmng_readraw(nandtool_common.oid, addr, buf, pagesz);
+		}
+		else {
+			ret = read(nandtool_common.fd, buf, pagesz);
+		}
+
+		if (ret != pagesz) {
+			fprintf(stderr, "nandtool: Fail to read a block at offset: %x\n", addr);
+			break;
+		}
+
+		if (write(outfd, buf, pagesz) < 0) {
+			fprintf(stderr, "nandtool: Fail to write to output file\n");
```

Comment:
strerror.

Also:
- detect partial write as an erorr (so `write() != ret`)
- You can name the variable better as it's only used for storing the read/write length